browser/net/brave_proxying_url_loader_factory.cc

```diff
@@ -0,0 +1,969 @@
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/net/brave_proxying_url_loader_factory.h"
+
+#include <utility>
+
+#include "base/bind.h"
+#include "base/debug/alias.h"
+#include "base/debug/dump_without_crashing.h"
+#include "base/feature_list.h"
+#include "base/strings/stringprintf.h"
+#include "base/task/post_task.h"
+#include "brave/browser/net/brave_request_handler.h"
+#include "brave/components/brave_shields/browser/adblock_stub_response.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/browser_task_traits.h"
+#include "content/public/browser/render_frame_host.h"
+#include "content/public/browser/resource_context.h"
+#include "content/public/common/url_utils.h"
+#include "mojo/public/cpp/system/string_data_pipe_producer.h"
+#include "net/base/completion_repeating_callback.h"
+#include "net/http/http_util.h"
+#include "services/network/public/cpp/features.h"
+#include "third_party/blink/public/platform/resource_request_blocked_reason.h"
+#include "url/origin.h"
+
+namespace {
+
+// User data key for ResourceContextData.
+const void* const kResourceContextUserDataKey = &kResourceContextUserDataKey;
+
+class ResourceContextData : public base::SupportsUserData::Data {
+ public:
+  ~ResourceContextData() override {}
+
+  static void StartProxying(
+      content::ResourceContext* resource_context,
+      int render_process_id,
+      int frame_tree_node_id,
+      network::mojom::URLLoaderFactoryRequest request,
+      network::mojom::URLLoaderFactoryPtrInfo target_factory,
+      network::mojom::TrustedURLLoaderHeaderClientRequest
+          header_client_request) {
+    DCHECK_CURRENTLY_ON(content::BrowserThread::IO);
+
+    auto* self = static_cast<ResourceContextData*>(
+        resource_context->GetUserData(kResourceContextUserDataKey));
+    if (!self) {
+      self = new ResourceContextData();
+      resource_context->SetUserData(kResourceContextUserDataKey,
+                                    base::WrapUnique(self));
+    }
+
+    if (!self->request_handler_) {
+      self->request_handler_.reset(new BraveRequestHandler);
+    }
+
+    auto proxy = std::make_unique<BraveProxyingURLLoaderFactory>(
+        self->request_handler_.get(), resource_context, render_process_id,
+        frame_tree_node_id, std::move(request), std::move(target_factory),
+        std::move(header_client_request), self->request_id_generator_,
+        base::BindOnce(&ResourceContextData::RemoveProxy,
+                       self->weak_factory_.GetWeakPtr()));
+
+    self->proxies_.emplace(std::move(proxy));
+  }
+
+  void RemoveProxy(BraveProxyingURLLoaderFactory* proxy) {
+    auto it = proxies_.find(proxy);
+    DCHECK(it != proxies_.end());
+    proxies_.erase(it);
+  }
+
+ private:
+  ResourceContextData()
+      : request_id_generator_(base::MakeRefCounted<RequestIDGenerator>()),
+        weak_factory_(this) {}
+
+  std::unique_ptr<BraveRequestHandler> request_handler_;
```

Comment:
We maintain one `ResourceContextData` per profile. It owns `BraveRequestHandler` and all factories. Also provides factories with id generator